File: library/src/main/java/me/proxer/library/api/AdaptionInfoAdapter.java
package me.proxer.library.api;

import com.squareup.moshi.FromJson;
import com.squareup.moshi.Json;
import me.proxer.library.entity.info.AdaptionInfo;
import me.proxer.library.enums.Medium;
import me.proxer.library.util.ProxerUtils;

/**
 * @author Ruben Gees
 */
class AdaptionInfoAdapter {

    @FromJson
    AdaptionInfo fromJson(final IntermediateAdaptionInfo json) {
        return new AdaptionInfo(json.id, json.name, ProxerUtils.toApiEnum(Medium.class, json.medium));
    }

    static class IntermediateAdaptionInfo {

        @Json(name = "id")
        String id;

        @Json(name = "name")
        String name;

        @Json(name = "medium")
        String medium;
    }
}

File: library/src/main/kotlin/me/proxer/library/internal/adapter/AdaptionInfoAdapter.kt
package me.proxer.library.internal.adapter

import com.squareup.moshi.FromJson
import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass
import me.proxer.library.entity.info.AdaptionInfo
import me.proxer.library.enums.Medium
import me.proxer.library.util.ProxerUtils

/**
 * @author Ruben Gees
 */
internal class AdaptionInfoAdapter {

    @FromJson
    fun fromJson(json: IntermediateAdaptionInfo): AdaptionInfo {
        return AdaptionInfo(json.id, json.name, ProxerUtils.toApiEnum<Medium>(json.medium))
    }

    @JsonClass(generateAdapter = true)
    internal data class IntermediateAdaptionInfo(
        @Json(name = "id") var id: String,
        @Json(name = "name") var name: String,
        @Json(name = "medium") var medium: String
    )
}
